Remove USE_CUDA from c10d reducer/logger

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59563 Move c10d to libtorch(_cuda)
* #59697 Add guards for USE_C10D_FOO in relevant c10d files
* #59696 Fix Windows guards in c10d
* **#59562 Remove USE_CUDA from c10d reducer/logger**
* #59561 Remove USE_CUDA from ProcessGroupGloo

Differential Revision: [D28931378](https://our.internmc.facebook.com/intern/diff/D28931378/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28931378/)!

lw committed:
Remove USE_CUDA from c10d reducer/logger



Differential Revision: [D28931378](https://our.internmc.facebook.com/intern/diff/D28931378/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28931378/)!

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from c10d reducer/logger"


Differential Revision: [D28931378](https://our.internmc.facebook.com/intern/diff/D28931378/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28931378/)!

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from c10d reducer/logger"


Differential Revision: [D28931378](https://our.internmc.facebook.com/intern/diff/D28931378/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28931378/)!

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from c10d reducer/logger"


Differential Revision: [D28931378](https://our.internmc.facebook.com/intern/diff/D28931378/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28931378/)!

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from c10d reducer/logger"


Differential Revision: [D28931378](https://our.internmc.facebook.com/intern/diff/D28931378/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28931378/)!

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from c10d reducer/logger"


Differential Revision: [D28931378](https://our.internmc.facebook.com/intern/diff/D28931378/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28931378/)!

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from c10d reducer/logger"


Differential Revision: [D28931378](https://our.internmc.facebook.com/intern/diff/D28931378/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28931378/)!

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from c10d reducer/logger"


Differential Revision: [D28931378](https://our.internmc.facebook.com/intern/diff/D28931378/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28931378/)!

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from c10d reducer/logger"


Differential Revision: [D28931378](https://our.internmc.facebook.com/intern/diff/D28931378/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28931378/)!

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from c10d reducer/logger"


Differential Revision: [D28931378](https://our.internmc.facebook.com/intern/diff/D28931378/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28931378/)!

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from c10d reducer/logger"


Differential Revision: [D28931378](https://our.internmc.facebook.com/intern/diff/D28931378/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28931378/)!

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from c10d reducer/logger"


Differential Revision: [D28931378](https://our.internmc.facebook.com/intern/diff/D28931378/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28931378/)!

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from c10d reducer/logger"


Differential Revision: [D28931378](https://our.internmc.facebook.com/intern/diff/D28931378/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28931378/)!

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from c10d reducer/logger"


Differential Revision: [D28931378](https://our.internmc.facebook.com/intern/diff/D28931378/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28931378/)!

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from c10d reducer/logger"


Differential Revision: [D28931378](https://our.internmc.facebook.com/intern/diff/D28931378/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28931378/)!

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from c10d reducer/logger"


Differential Revision: [D28931378](https://our.internmc.facebook.com/intern/diff/D28931378/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28931378/)!

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from c10d reducer/logger"


Differential Revision: [D28931378](https://our.internmc.facebook.com/intern/diff/D28931378/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28931378/)!

[ghstack-poisoned]

zhaojuanmao reviewed (commented):
love the change, it makes codes much cleaner, thanks!

before landing, would you please run a benchmark for your PR and verify the measured times are correct?

lw committed:
Update on "Remove USE_CUDA from c10d reducer/logger"


Differential Revision: [D28931378](https://our.internmc.facebook.com/intern/diff/D28931378/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28931378/)!

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from c10d reducer/logger"


Differential Revision: [D28931378](https://our.internmc.facebook.com/intern/diff/D28931378/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28931378/)!

[ghstack-poisoned]

agolynski commented on torch/lib/c10d/reducer_cuda.cpp:
```diff
@@ -0,0 +1,91 @@
+        TORCH_INTERNAL_ASSERT(false);
+    }
+    // In case the compiler doesn't get it.
+    return forward_start;
```
could you get a warning/lint on unreachable code instead?

lw replied:
Dunno, it doesn't seem we're getting any such lints from clang-tidy though? But let me remove this if you prefer

agolynski commented on torch/lib/c10d/reducer_cuda.cpp:
```diff
@@ -0,0 +1,91 @@
+#include <c10d/reducer.hpp>
```
call this module cuda_timer?

i wonder if it's possible to implement just one timer (instead of CPU and GPU timers) with c10::Events?

lw replied:
The class is called `CudaTimer`. Unless you're talking about the file?

I chose to call the file `reducer_cuda` because it contains the CUDA-related stuff for the `reducer.cpp` file. If I saw a `timer_cuda.cpp` file I would expect there to be also a `timer.cpp` file (or `timer_cpu.cpp`) but that's not the case.

> i wonder if it's possible to implement just one timer (instead of CPU and GPU timers) with c10::Events?

There's several obstacles to this:
- c10::Event doesn't have an `elapsed_time` method (though we could add it but it's unclear how to make it device-agnostic)
- even if we did so, c10::Events of "CPU" type currently behave like no-ops, thus they're unable to provide timing information, which means we'd still need two separate branches for CPU and CUDA

agolynski replied:
Thanks.

Do you feel like it'd be worth while fixing c10::Event so it aligns with CUDAEvent to move further to "CUDA agnostic" code?

agolynski replied:
Regarding timer: do you feel that timer is only going to be used in the context of reducer, e.g. maybe timing collectives or even something related to RPC?

lw replied:
Let me ask @ezyang: do you think we could/should add an `elapsed_time` method to c10::Event? (like the one on CUDAEvent)

ezyang replied:
I don't think I have enough information right now to say. Under what situations would you exercise this code with CPU events?

lw replied:
@ezyang I don't think you can have c10::Events with a CPU device? (All event-related methods of the CPU DeviceGuardImpl throw). Which actually means that we'd still need two separate codepaths for CPU and non-CPU. My question was whether `elapsed_time` is a "generic enough" method to elevate to c10::Event.

Anyway, in the end I won't be doing this change here to avoid blocking the stack, I'll consider it for later.

lw replied:
> Regarding timer: do you feel that timer is only going to be used in the context of reducer, e.g. maybe timing collectives or even something related to RPC?

In its current version, Timer is very much specific to the reducer since it hardcodes the 5 exact types of events it can be used for (forward, backward_comm, ...). If we'll ever make Timer more generic we can move it to a better location at that time.

ezyang replied:
> My question was whether elapsed_time is a "generic enough" method to elevate to c10::Event.

OK. Well, the way I would think about this problem is that the more API surface you put on a virtual interface, the harder it is to bring up new implementations of the interface. This isn't a terribly pressing concern right now, because CUDA events are the only real game in town, but it would be more difficulty later. My general inclination is, as you've done here, keep things specialized if there is only one user, and let people who are later trying to generalize make the call.

agolynski reviewed (approved):
LG

lw committed:
Update on "Remove USE_CUDA from c10d reducer/logger"


Differential Revision: [D28931378](https://our.internmc.facebook.com/intern/diff/D28931378/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28931378/)!

[ghstack-poisoned]
